### PR TITLE
Run QC for WTS alongside WGS

### PR DESCRIPTION
**Pull request by alexiswl**
# cba747810c875ad9c4a919c16b62bdbb3cca6be8

> Updates by file

## data_processors.pipeline.domain.workflow.py

* Added abstract workflow type DRAGEN_WGTS_QC.
* Added labmetadata rule method must_be_wgts

## data_processors.pipeline.lambdas.dragen_wgs_qc.py

* Handle WTS data and add enable_rna option. True for WTS data, false for WGS data

## data_processors.pipeline.orchestartion.dragen_wgs_qc_step.py

* Set batcher runstep parameter to DRAGEN_WGTS_QC abstract workflow type
* Replace must_be_wgs labmetadata rule with must_be_wgts
* Update warning to use DRAGEN_WGTS_QC  abstract workflow type

# 5eb62f375ff69f5bb4066ce5b065afb62c262635

> Update orchestration logic

* Now dragen wts happens after QC is complete.

* Instead of performing dragen_wts after bclconvert, we first run dragen_wgts_qc (which now includes dragen_wts samples).

### CODE REVIEW

**Comment by alexiswl** [comment]
[Run WTS QC workflow through Portal before running Dragen WTS Tumor Only Workflow](https://trello.com/c/H0DQyo2B/1485-run-wts-qc-workflow-through-portal-before-running-dragen-wts-tumor-only-workflow)

File: data_processors/pipeline/lambdas/orchestrator.py [hunk 1]
```diff
@@ -223,9 +217,11 @@ def next_step(this_workflow: Workflow, skip: dict, context=None):
 
         return results
 
-    elif this_workflow.type_name.lower() == WorkflowType.DRAGEN_WGS_QC.value.lower() and \
-            this_workflow.end_status.lower() == WorkflowStatus.SUCCEEDED.value.lower():
-        logger.info(f"Received DRAGEN_WGS_QC workflow notification")
+    elif (
+            this_workflow.type_name.lower() == WorkflowType.DRAGEN_WGS_QC.value.lower() or
+            this_workflow.type_name.lower() == WorkflowType.DRAGEN_WTS_QC.value.lower()
+    ) and this_workflow.end_status.lower() == WorkflowStatus.SUCCEEDED.value.lower():
```
Comment (victorskl) [inline data_processors/pipeline/lambdas/orchestrator.py:223]: I still wish to keep this just WGS check here then triggering T/N. More in below comment.

File: data_processors/pipeline/lambdas/orchestrator.py [hunk 2]
```diff
@@ -243,6 +239,12 @@ def next_step(this_workflow: Workflow, skip: dict, context=None):
             logger.info("Performing TUMOR_NORMAL_STEP")
             results.append(tumor_normal_step.perform(this_workflow))
 
+        if "DRAGEN_WTS_STEP" in skiplist:
+            logger.info("Skip performing DRAGEN_WTS_STEP")
+        else:
+            logger.info("Performing DRAGEN_WTS_STEP")
+            results.append(dragen_wts_step.perform(this_workflow))
```
Comment (victorskl) [inline data_processors/pipeline/lambdas/orchestrator.py:246]: At the mo, WTS (transcriptome) are still triggering in Batch fashion. See around [1] for example.

So this will break that part a bit. Because, within this if block, T/N will be detached from Batch view and, trigger by Subject across sequence runs basis then on.

WTS is still per library within this Batch/sequence run.

So. We should better keep these different impedance mis-match, separated with their own if/else block. 

Furthermore, since `wts_alignment_qc` are in their own Batch run context, so we have to better check all in this QC Batch run come clean. Only then we can start of triggering next WTS (transcriptome) Batch run. This logic may be better implemented in step module side i.e. part of pre-step in `dragen_wts_step.preform()`.


[1]: https://umccr.slack.com/archives/C8CG6K76W/p1685863962967909

**Review by victorskl — COMMENTED** [review]
Looking good most of it, Alexis. Just that need to attend comment on Orchestrator on triggering cue.

File: data_processors/pipeline/lambdas/orchestrator.py [hunk 3]
```diff
@@ -187,10 +187,10 @@ def next_step(this_workflow: Workflow, skip: dict, context=None):
             logger.info("Updating Google LIMS")
             google_lims_update_step.perform(this_workflow)
 
-        if "DRAGEN_WGS_QC_STEP" in skiplist:
-            logger.info("Skip performing DRAGEN_WGS_QC_STEP")
+        if "DRAGEN_WGTS_QC_STEP" in skiplist:
```
Comment (victorskl) [inline data_processors/pipeline/lambdas/orchestrator.py:190]: Can you also update this new type in `handler()` event payload dict docstring, pls.

File: data_processors/pipeline/lambdas/orchestrator.py (hunk 2 above)
Comment (alexiswl) [reply data_processors/pipeline/lambdas/orchestrator.py:246]: > WTS is still per library within this Batch/sequence run.

Ah okay, I think we should switch to the same logic as TN workflow. If a dragen WTS QC workflow is launched per lane we want to wait until all lanes for that library are completed before launching the WTS workflow right?

Comment (alexiswl) [reply data_processors/pipeline/lambdas/orchestrator.py:246]: Hence rather than having the Batcher class as input into the prepare_dragen_wts_jobs step we should instead have a List[LabMetadata] object instead?

File: data_processors/pipeline/domain/workflow.py [hunk 4]
```diff
@@ -29,7 +29,9 @@ class SampleSheetCSV(Enum):
 
 class WorkflowType(Enum):
     BCL_CONVERT = "bcl_convert"
+    DRAGEN_WGTS_QC = "wgts_alignment_qc"  # Placeholder workflow type
     DRAGEN_WGS_QC = "wgs_alignment_qc"
+    DRAGEN_WTS_QC = "wts_alignment_qc"
```
[the remark this answers is not in the document]
Comment (alexiswl) [reply data_processors/pipeline/domain/workflow.py:34]: Resolved in 1186883cb382396e469671b2dc8b2c768c39d448

File: data_processors/pipeline/lambdas/dragen_wgs_qc.py [hunk 5]
```diff
@@ -1,3 +1,5 @@
+from data_portal.models.labmetadata import LabMetadataType
```
[the remark this answers is not in the document]
Comment (alexiswl) [reply data_processors/pipeline/lambdas/dragen_wgs_qc.py]: Resolved in 1186883cb382396e469671b2dc8b2c768c39d448

File: data_processors/pipeline/lambdas/orchestrator.py (hunk 1 above)
Comment (alexiswl) [reply data_processors/pipeline/lambdas/orchestrator.py]: Partially rolled-back in 1186883cb382396e469671b2dc8b2c768c39d448

File: data_processors/pipeline/lambdas/orchestrator.py (hunk 3 above)
Comment (alexiswl) [reply data_processors/pipeline/lambdas/orchestrator.py:190]: Resolved in 1186883cb382396e469671b2dc8b2c768c39d448

File: data_processors/pipeline/lambdas/orchestrator.py (hunk 2 above)
Comment (victorskl) [reply data_processors/pipeline/lambdas/orchestrator.py:251]: > I think we should switch to the same logic as TN workflow.

Yup. Agree. It seems business logic/requirement is changed for triggering transcriptome workflow.

> If a dragen WTS QC workflow is launched per lane we want to wait until all lanes for that library are completed before launching the WTS workflow right?

Correct. More below.

> Hence rather than having the Batcher class as input into the `prepare_dragen_wts_jobs` step we should instead have a List[LabMetadata] object instead?

Yep. WTS becomes very much similar to T/N situation. But. Instead of pooling across sequencing runs, it has to find its peer _(like "minions", ya know!)_ that might spread across different lanes within this sequencing batch.

So, we can -- 
* Remove the `Batcher -> Batch/BatchRun` mechanism wrap from `dragen_wts_step`
* The input `this_workflow` of its interface `dragen_wts_step.perform(this_workflow)` is no longer `bcl_convert` workflow anymore
* So we loose `FastqListRows` context along with `Batcher` there
* However, `this_workflow` which is now `type_name == wts_alignment_qc` of WTS lib in per lane has still a link to `Workflow.sequence_run_id` attribute
* With that we can still scope it into the same sequencing run context
  * This will then give ability to reconstruct related `LibraryRun` 
  * And/or as well as all the way back to its parent workflow `type_name == bcl_convert`
  * _(...can use one or both of them as find fit)_ ☝️ 
* Then, we can limit the searching space for finding WTS library its other half ("minions") that went into another lanes

Comment (alexiswl) [reply data_processors/pipeline/lambdas/orchestrator.py:251]: I've updated this in a552dccc49fd1a5af04542f541f1ba00cf2ecbe0, would you mind having a look? 

If that's all good I'll fix up the tests (assuming I've broken a few with these changes)

File: data_processors/pipeline/orchestration/dragen_wts_step.py [hunk 6]
```diff
@@ -26,90 +26,99 @@
...
-        workflow=this_workflow,
-        run_step=WorkflowType.DRAGEN_WTS.value,
-        batch_srv=batch_srv,
-        fastq_srv=fastq_srv,
-        logger=logger,
+    this_sqr = this_workflow.sequence_run
+
+    # Check if all other DRAGEN_WTS_QC workflows for this run have finished
+    # If yes we continue to the DRAGEN WTS Workflow
+    # If not we wait (until all DRAGEN_WTS_QC worklfows for this run have finished
```
Comment (victorskl) [inline data_processors/pipeline/orchestration/dragen_wts_step.py:33]: typo in comment

```
workflows
```

File: data_processors/pipeline/orchestration/dragen_wts_step.py [hunk 7]
```diff
@@ -26,90 +26,99 @@
...
+
+        if job_list:
+            logger.info(f"Submitting {len(job_list)} WTS jobs for {submitting_subjects}.")
+            queue_arn = libssm.get_ssm_param(SQS_DRAGEN_WTS_QUEUE_ARN)
+            libsqs.dispatch_jobs(queue_arn=queue_arn, job_list=job_list)
+        else:
+            logger.warning(f"Calling to prepare_tumor_normal_jobs() return empty list, no job to dispatch...")
     else:
-        batcher.reset_batch_run()  # reset running if job_list is empty
+        logger.warning(f"DRAGEN_WGS_QC workflow finished, but {len(running)} still running. Wait for them to finish...")
```
Comment (victorskl) [inline data_processors/pipeline/orchestration/dragen_wts_step.py:61]: typo in logger

```
DRAGEN_WTS_QC
```

File: data_processors/pipeline/orchestration/tumor_normal_step.py [hunk 8]
```diff
@@ -19,6 +19,7 @@
     _mint_libraries
 from data_processors.pipeline.services import workflow_srv, metadata_srv, fastq_srv
 from data_processors.pipeline.tools import liborca
+from data_processors.pipeline.tools.liborca import _handle_rerun
```
Comment (victorskl) [inline data_processors/pipeline/orchestration/tumor_normal_step.py:22]: Could put this under `__init__.py` of `data_processors.pipeline.orchestration` package, pls. Along with other underscore methods there.

```
data_processors.pipeline.orchestration._handle_rerun(...)
```

**Review by victorskl — COMMENTED** [review]
LGTM, few more minor comments to attend.

File: data_processors/pipeline/lambdas/orchestrator.py (hunk 2 above)
Comment (victorskl) [reply data_processors/pipeline/lambdas/orchestrator.py:251]: Looking so far, Alexis. Added few more comments to attend.

File: data_processors/pipeline/orchestration/dragen_wts_step.py (hunk 7 above)
Comment (alexiswl) [reply data_processors/pipeline/orchestration/dragen_wts_step.py]: Resolved in 7f4f03c68c8146c4034f141bf454a5ff7b4f5321

File: data_processors/pipeline/orchestration/dragen_wts_step.py (hunk 6 above)
Comment (alexiswl) [reply data_processors/pipeline/orchestration/dragen_wts_step.py]: Resolved in 7f4f03c68c8146c4034f141bf454a5ff7b4f5321

File: data_processors/pipeline/orchestration/tumor_normal_step.py (hunk 8 above)
Comment (alexiswl) [reply data_processors/pipeline/orchestration/tumor_normal_step.py]: Resolved in 7f4f03c68c8146c4034f141bf454a5ff7b4f5321

File: data_processors/pipeline/orchestration/__init__.py [hunk 9]
```diff
@@ -25,9 +25,13 @@
 
 import pandas as pd
 
+from data_portal.models.fastqlistrow import FastqListRow
 from data_portal.models.labmetadata import LabMetadata, LabMetadataPhenotype, LabMetadataType
 from data_processors.pipeline.tools import liborca
 
+logger = logging.getLogger(__name__)
```
Comment (victorskl) [inline data_processors/pipeline/orchestration/__init__.py:32]: Need `import logging`

**Review by victorskl — COMMENTED** [review]
Hey Alexis;

1)
Could you please change PR to ready for review from draft, if you are ready..? This looks good so far from my side, minor comments to attend.


2)
> If that's all good I'll fix up the tests (assuming I've broken a few with these changes)

Do you still have bandwidth to follow up tests..?
If so, few pointers as follows...

- Each python module file that has made changes; should have their test file counterpart in `tests` package.
- First, you run each test file for unit test cases and fix where it broke, accordingly.
- Optionally, run integration test cases there, if applicable and/or see fit.
- Finally, run smoke test `make test` on the whole suite and, it must pass.

Here are starter, based on file changes in PR.

- Test domain logic code changes
```
python manage.py test data_processors.pipeline.domain.tests.test_workflow
```

- Test front controller (Lambda) entrypoints changes
```
python manage.py test data_processors.pipeline.lambdas.tests.test_dragen_wgs_qc
python manage.py test data_processors.pipeline.lambdas.tests.test_dragen_wts
python manage.py test data_processors.pipeline.lambdas.tests.test_orchestrator
```

- Test orchestration step module middleware changes
```
python manage.py test data_processors.pipeline.orchestration.tests.test_dragen_wgs_qc_step
python manage.py test data_processors.pipeline.orchestration.tests.test_dragen_wts_step
python manage.py test data_processors.pipeline.orchestration.tests.test_tumor_normal_step
```

- Test service layer module changes
```
python manage.py test data_processors.pipeline.services.tests.test_metadata_srv
```

**Comment by alexiswl** [comment]
Thanks Victor, I'll work on this tomorrow if you don't mind?

**Comment by victorskl** [comment]
Of course, go for it, please..!

File: data_processors/pipeline/lambdas/dragen_wgs_qc.py [hunk 10]
```diff
@@ -103,13 +104,28 @@ def handler(event, context) -> dict:
 
     sample_name = fastq_list_rows[0]['rgsm']
 
-    # Set workflow helper
-    wfl_helper = SecondaryAnalysisHelper(WorkflowType.DRAGEN_WGS_QC)
+    # Get metadata by library id
+    library_lab_metadata = metadata_srv.get_metadata_by_library_id(library_id)
+
+    # We set the RNA flag and set the workflow type based on the library lab metadata
+    if library_lab_metadata.type == LabMetadataType.WTS:
```
Comment (victorskl) [inline data_processors/pipeline/lambdas/dragen_wgs_qc.py:111]: As discussed. The `metadata_srv.get_metadata_by_library_id(library_id)` could return `None` if not found in database.

Hence. We should better handle `if library_lab_metadata is None` before accessing its attribute `library_lab_metadata.type` value checking.

File: data_processors/pipeline/lambdas/orchestrator.py [hunk 11]
```diff
@@ -242,6 +236,19 @@ def next_step(this_workflow: Workflow, skip: dict, context=None):
         else:
             logger.info("Performing TUMOR_NORMAL_STEP")
             results.append(tumor_normal_step.perform(this_workflow))
+    elif this_workflow.type_name.lower() == WorkflowType.DRAGEN_WTS_QC.value.lower() and \
```
[the remark this answers is not in the document]
Comment (alexiswl) [reply data_processors/pipeline/lambdas/orchestrator.py:241]: Resolved in 
55642c0

File: data_processors/pipeline/lambdas/dragen_wgs_qc.py (hunk 10 above)
Comment (alexiswl) [reply data_processors/pipeline/lambdas/dragen_wgs_qc.py:116]: Resolved in 7583771

File: data_processors/pipeline/lambdas/dragen_wgs_qc.py [hunk 12]
```diff
@@ -103,13 +104,28 @@ def handler(event, context) -> dict:
...
+        workflow_type = WorkflowType.DRAGEN_WTS_QC
+        enable_rna = True
+    elif library_lab_metadata.type == LabMetadataType.WGS:
+        workflow_type = WorkflowType.DRAGEN_WGS_QC
+        enable_rna = False
+    else:
+        logger.error(f"Expected metadata type for library id '{library_id}' to be one of WGS or WTS")
+        raise ValueError
+
+    wfl_helper = SecondaryAnalysisHelper(workflow_type)
```
[the remark this answers is not in the document]
Comment (alexiswl) [reply data_processors/pipeline/lambdas/dragen_wgs_qc.py:126]: Resolved in 55642c0

**Comment by alexiswl** [comment]
Getting close!

```
======================================================================
ERROR: test_perform (data_processors.pipeline.orchestration.tests.test_dragen_wts_step.DragenWtsStepUnitTests.test_perform)
python manage.py test data_processors.pipeline.orchestration.tests.test_dragen_wts_step.DragenWtsStepUnitTests.test_perform
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/alexiswl/UMCCR/GitHub/data-portal-apis/data_processors/pipeline/orchestration/tests/test_dragen_wts_step.py", line 95, in test_perform
    self.assertEqual(results['submitting_subjects'][0], wts_mock_subject_id)
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range

======================================================================
ERROR: test_perform (data_processors.pipeline.orchestration.tests.test_somalier_extract_step.SomalierExtractStepUnitTests.test_perform)
python manage.py test data_processors.pipeline.orchestration.tests.test_somalier_extract_step.SomalierExtractStepUnitTests.test_perform
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/alexiswl/UMCCR/GitHub/data-portal-apis/data_processors/pipeline/orchestration/tests/test_somalier_extract_step.py", line 88, in test_perform
    self.assertIn('index', results['somalier_extract_step'][0])
                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range

----------------------------------------------------------------------
Ran 315 tests in 46.693s

FAILED (errors=2, skipped=40)
Destroying test database for alias 'default'...
make: *** [Makefile:65: test] Error 1
```

**Comment by alexiswl** [comment]
@victorskl would you mind having a look at these? I couldn't figure out the changes needed to get the test perform working and I think the orchestration logic might not be handing the DRAGEN_WGTS_QC step?

**Comment by victorskl** [comment]
Ok, I will pull and check in tick.

**Comment by victorskl** [comment]
Hi Alexis; pls see https://github.com/umccr/data-portal-apis/pull/607 for 
```
python manage.py test data_processors.pipeline.orchestration.tests.test_dragen_wts_step.DragenWtsStepUnitTests.test_perform
```

TL;DR - it just need mock WTS tumor Lab meta info <> LibraryRun <> Workflow has to link up / match them up well for test condition.

**Comment by victorskl** [comment]
I will check for somalier step, next..

**Comment by victorskl** [comment]
Ah this one is easy

```
python manage.py test data_processors.pipeline.orchestration.tests.test_somalier_extract_step.SomalierExtractStepUnitTests.test_perform
```

Just roll back changes here

https://github.com/umccr/data-portal-apis/blob/55642c0a8c9bf55d75ff1781cde6abf0033be33e/data_processors/pipeline/orchestration/tests/test_somalier_extract_step.py#L67

... to

```
mock_wgc_qc_workflow: Workflow = DragenWgsQcWorkflowFactory()
```

---

Explain

- We don't really save to db with a workflow type `WorkflowType.DRAGEN_WGTS_QC` yet
- It is still a "placeholder" or "common" workflow type to process together both WGS/WTS alignment QC
- So, we should just still keep this unittest case as is. Which is, originally unit testing `WorkflowType.DRAGEN_WGS_QC` scenario and, still valid test case.

**Comment by alexiswl** [comment]
Woo!

```
...
----------------------------------------------------------------------
Ran 315 tests in 48.655s

OK (skipped=40)
Destroying test database for alias 'default'...
```

**Review by victorskl — APPROVED** [review]
Go for it, Alexis..!
